source/common/router/router.cc

```diff
@@ -70,12 +71,30 @@ bool convertRequestHeadersForInternalRedirect(Http::HeaderMap& downstream_header
     return false;
   }
 
+  // Don't allow serving TLS responses over plaintext.
   bool scheme_is_http = schemeIsHttp(downstream_headers, connection);
   if (scheme_is_http && absolute_url.scheme() == Http::Headers::get().SchemeValues.Https) {
-    // Don't allow serving TLS responses over plaintext.
     return false;
   }
 
+  // Make sure that performing the redirect won't result in exceeding the configured number of
+  // redirects allowed for this route.
+  if (!filter_state.hasData<StreamInfo::UInt32Accessor>(
+          NumPreviousInternalRedirectFilterStateName)) {
+    filter_state.setData(NumPreviousInternalRedirectFilterStateName,
+                         std::make_shared<StreamInfo::UInt32AccessorImpl>(0),
+                         StreamInfo::FilterState::StateType::Mutable,
+                         StreamInfo::FilterState::LifeSpan::DownstreamRequest);
+  }
+  StreamInfo::UInt32Accessor& num_previous_internal_redirect =
+      filter_state.getDataMutable<StreamInfo::UInt32Accessor>(
+          NumPreviousInternalRedirectFilterStateName);
+
+  if (num_previous_internal_redirect.value() >= max_internal_redirects) {
```

Comment:
What do you think about renaming this as well?

Reply:
Yep. Done. I kept the helper method in the test to still be setNumPreviousRedirect because I think in that context, it is proper to say that "this is to set up the test filter to test the case when it sees there are x simulated previous internal redirects".